### ISSUE TITLE
Riot.im/app sends a GET request to matrix.org while logged in to another homeserver

### ISSUE THREAD

**Issue by aaronraimist** [vector-im/riot-web#11655]
riot.im/app and /develop send a GET request to https://matrix-client.matrix.org/_matrix/client/versions and https://matrix.org/.well-known/matrix/client even when you are logged in to another homeserver.

**Comment by turt2live** [comment]
This is to ensure the config is valid before you embark on a journey through Riot. It's possible your homeserver logs you out, or you log yourself out, in which case the auth pages should still work.

**Comment by turt2live** [comment]
This gets critical status because it affects the vast majority of our users.

**Comment by sneak** [comment]
Note that this bug's description is scoped too narrowly: it appears to send the requests on *first launch*, when you've never logged in anyway.  (These statements are based on @turt2live's assertion that #12712 is a dupe - that one is about first-launch, never-logged-in behavior.)

**Comment by turt2live** [comment]
This issue is about that too: the config is validated on every startup.

**Comment by sneak** [comment]
Sure, but if the default config is for `matrix.org`, then "config validation" is a form of accidental telemetry.  Is it absolutely required that config validation hit the network?

**Comment by turt2live** [comment]
That's why this is also `needs-product-decision`. It is categorically not intended to be telemetry, it's just to make sure the app works with whatever build config is shipped with the app.

**Comment by sneak** [comment]
> It is categorically not intended to be telemetry

I appreciate that.  The military intelligence spies don't, and slurp it up anyway. :)

Thank you for prioritizing this!  Lots of projects don't take these sorts of leaks seriously.  On behalf of people who need privacy even more than I do, thank you!

**Comment by sneak** [comment]
Per #13942, it also connects to vector.im and riot.im, for a total of *three* unauthorized/unnecessary connections, even when just launched and sitting at the login/signup screen.  I've filed this as a separate issue, because it seems that this one is scoped to contacting matrix.org, when in reality there are *3* different privacy leaks.

It should be connecting only to the homeserver to which it is configured, and then only on signup/login.  Any other third party connections (such as vector.im, or an upgrade check if that's why it's contacting riot.im) should be opt in.

I don't want my privacy client talking to any services other than my self-hosted one, thank you.  This has been keeping me from using or recommending Riot/Matrix, because this is a major privacy leak.

**Comment by t3chguy** [comment]
As a product requirement, riot first thing it does as part of boot ensures that its config.json is valid, which means confirming all the default services specified there. 

For riot desktop you can change the config as per the docs. https://github.com/vector-im/riot-desktop#user-specified-configjson

**Comment by sneak** [comment]
Why are there any product requirements for network access before the user has instructed the client to do anything, sitting at the login screen?

**Comment by t3chguy** [comment]
Well you can't show the login screen without knowing whether the default service as specified in the config.json is valid.

**Comment by t3chguy** [comment]
For instance if the default HS as configured in the config.json is SSO-only then showing a username&password prompt is inappropriate.

**Comment by sneak** [comment]
If that's the policy, I think it's unreasonable in that case for the default config to assume that I want to use those three third-party services without asking me first.

**Comment by t3chguy** [comment]
Well that's up to you, depending on what Riot instance you use, if you download the riot-desktop build then that ships with a config that specifies those services. Any instance/distribution is free to provide config pointing at their own services.

**Comment by sneak** [comment]
Is that a polite way of telling me "fork it if you don't like it to connect to servers you don't want to talk to automatically on startup without notice or consent?"

**Comment by t3chguy** [comment]
You don't need to fork it at all, just specify your own config.json as I said earlier.

**Comment by sneak** [comment]
I'm talking about the product being a de-facto client for matrix.org, not my usage.  I can rip all of this telemetry out of the client and rebuild it if I wish; I'm talking about the potentially thousands or millions of people who download this client and expect to be able to use it to privately communicate with a non-matrix.org homeserver.  They can't do so privately.

**Comment by t3chguy** [comment]
How would you propose it show a login screen without knowing what login flows the config server uses? SSO? Username & Password? Email & Password?

**Comment by sneak** [comment]
Well, if the config is non-default, that seems like a reasonable approach, as the user or their distributor has explicitly configured things.

In the 99.9% case where a user has simply downloaded the app from the website and double-clicked it, it shouldn't need to do anything at that moment of launch.  If the user takes an action, any action, to attempt to use (log in, sign up) on `matrix.org`, then it should talk to `matrix.org`, but not before.  It should then prompt them to opt-in to using the `vector.im` lookup service at that time.

I assume (without any research) the `riot.im` connection is an autoupdate check?  That should also be opt in (perhaps with a modal, as annoying as they are, and perhaps after the client is open for a period of time as to not spam them on startup), as it serves as nonconsensual telemetry if the user doesn't want the `riot.im` operators or hosting provider to know their IP is using matrix/riot.

In the other case where someone downloads a client and wishes to use it with an alternate homeserver, it should not automatically assume the `matrix.org`/`vector.im` configuration values on fresh/clean startup.  A user action should happen before those values, and thus their network checks, occur.  That way a user can download the app, launch it, configure their custom homeserver in the GUI, and nobody from matrix.org, vector.im, or riot.im ever knows it happened.

**Comment by t3chguy** [comment]
Right but as soon as the user clicks `Login` the default server as per their config is matrix.org so without changing the flow to add yet another button click it'd have to already make a web request to know the flows to use there.

The update endpoint as per the default riot-desktop config is on packages.riot.im - https://github.com/vector-im/riot-desktop/blob/develop/riot.im/release/config.json#L2

> That should also be opt in (perhaps with a modal, as annoying as they are, and perhaps after the client is open for a period of time as to not spam them on startup)

They are opt-in, you opted into it by choosing the official distribution which has it always enabled. Bugs with out of date versions cannot be resolved.

Using the custom config.json I shared document to earlier you can also disable update checking.

> In the other case where someone downloads a client and wishes to use it with an alternate homeserver, it should not automatically assume the matrix.org/vector.im configuration values on fresh/clean startup. A user action should happen before those values, and thus their network checks, occur. That way a user can download the app, launch it, configure their custom homeserver in the GUI, and nobody from matrix.org, vector.im, or riot.im ever knows it happened.

Sadly a large number of users use matrix.org and adding an additional click to everyone to accept the defaults they chose by choosing the source they got their app from would be a really poor onboarding UX.

**Comment by sneak** [comment]
Ok, then I've got my answer.  Riot is an official client of the service offered by `matrix.org`, and I cannot recommend use of Riot to people with which I want to communicate for reasons mentioned above, because it will immediately connect to `matrix.org` and other supporting services on startup by default regardless of whether or not the end user is taking advantage of federation or not.

For me and my group of mostly nontechnical users, that makes the privacy leak serve as a nonstarter.  I'm very sad about that.

**Comment by t3chguy** [comment]
The current behaviour which came from product is to make sure the config.json is valid and nothing there is wrong, which means verifying all provided services and refusing to start the app if the config.json proves invalid. I will raise your concerns with the product team.

Overhauling the Login process to add another stage before the Login/Registration field e.g

1. Click Login
1. Click accept default homeserver
1. Enter credentials

(where 2 is new) is very intrusive to the majority of users which make use of their config.json specified default servers.

**Comment by t3chguy** [comment]
> Ok, then I've got my answer. Riot is an official client of the service offered by matrix.org

I am not speaking on behalf of the product team, am only reciting what I have heard in the past.

**Comment by t3chguy** [comment]
You may be misunderstanding me, I am not going against privacy, I am all for it. I am trying to explain the decisions that led to the current behaviour and offer reasonable workarounds by means of the config.json

I have raised it to be flagged to Product with higher priority as https://github.com/vector-im/riot-web/issues/11655#issuecomment-598273959 should have triggered it to be

**Comment by t3chguy** [comment]
The main symptom issues here are:

+ Don't connect to default server if you are logged into another server
+ Don't connect to default server until the user explicitly approves it
+ Don't assume the user wants auto-updating until the user explicitly approves it (most users click away on any such asks so would end up never updating)

**Comment by sneak** [comment]
I don't understand the need to communicate with any server *prior* to the user clicking "sign up" or "log in", because they are not performing a registration or a login prior to that point when sitting at the splash screen.  It is only at the point of clicking one of those buttons that it needs to communicate with (any) homeserver.

The change could be as simple as adding a "Specify custom homeserver..." link, or perhaps having a drop-down above the "sign in" / "create account" screen labeled "Homeserver" with a default setting of "matrix.org" and a second item that says "custom..." allowing the user to input their own homeserver (or select from other community-run open registration ones other than `matrix.org`).

This bug can be summarized alternately as "the launch splash screen does not need the network to render itself".

**Comment by sneak** [comment]
> Don't assume the user wants auto-updating until the user explicitly approves it (most users click away on any such asks so would end up never updating)

You can make the default option "yes" so when they "click away" it offers autoupdate.

Alternately, you can make the autoupdate check happen a random number of minutes after startup, so it doesn't serve as a naked "application launched" telemetry event.

**Comment by t3chguy** [comment]
> I don't understand the need to communicate with any server prior to the user clicking "sign up" or "log in", because they are not performing a registration or a login prior to that point when sitting at the splash screen. It is only at the point of clicking one of those buttons that it needs to communicate with (any) homeserver.

The explanation is here: https://github.com/vector-im/riot-web/issues/11655#issuecomment-565077129

I don't personally understand that decision either, hence me planning to raise it with Product.

> Alternately, you can make the autoupdate check happen a random number of minutes after startup

It shouldn't happen at startup, it happens 30s after startup: https://github.com/vector-im/riot-desktop/blob/develop/src/updater.js#L4 and then subsequently every hour

**Comment by t3chguy** [comment]
> You can make the default option "yes" so when they "click away" (interpreted as smashing the default button on the modal) it enables the autoupdate.

Sure, there's quite a lot of things the user is already asked during onboarding and adding even another pop-up might be a deterrent.

**Comment by sneak** [comment]
Adding silent, unwanted third-party network requests without an opt-in is a hard deterrent to those that just want to use a piece of client software to talk to a server within their own organization without informing third parties that they are doing so.

Terms like "official distribution" seem to be implying that those people should be repackaging the client with a custom config file with telemetry and autoupdate disabled.  That's a bummer, because it means those users will either be out of date (I'm not going to run my own autoupdate api server) or will just skip using the Matrix protocol, which is what has happened to me.

**Comment by t3chguy** [comment]
Well orgs tend to have tools like GPO and such to deploy org-wide config which is designed for things like this without needing to repackage and that way wouldn't need to build anything themselves nor run their own autoupdater

**Comment by bpcurse** [comment]
As I already mentioned in #11648, please consider the legal implications (EU-GDPR) in this product decision, too. "Privacy by design/default"

**Comment by jryans** [comment]
Thanks for the feedback on this. I have collected these concerns, and I am discussing them with the Product team. I'll provide an update here once we've worked out what path Riot should take on this.

**Comment by jryans** [comment]
Thanks for your patience. This is a somewhat nuanced issue, as there are multiple different requests at or just after startup, and they are triggered for different reasons. We have discussed these concerns as a Product team, so I'll share our plans for each type of request as well.

### **1. Validate default homeserver and identity server from app config**

*   `GET https://matrix.org/.well-known/matrix/client`
*   `GET https://vector.im/_matrix/identity/api/v1`

On every app startup (even if you are already logged in), Riot currently validates the default homeserver and identity server for several reasons:

*   to assist Riot admins who might have made a typo in their config or misconfigured something server side
*   to ensure we can create a guest account for the welcome page ([slated for removal](https://github.com/vector-im/riot-web/issues/11533#issuecomment-559042353) as part of FTUE work)
*   to ensure email addresses for new accounts can be bound to a working identity server (less important after 2019 privacy project, since the homeserver now handles email for account management)
*   to ensure that if the user gets logged out for whatever reason (and happen to use the default homeserver) that they’ll theoretically be able to get back in

For this section, we plan to:

* Stop validating the homeserver and identity server at startup
* Remove room directory button from welcome page
* Defer guest account creation until the last possible moment when actually needed

The end result would be that these requests are no longer made during startup. We don't have bandwidth to take on this work right away, but I would expect it to begin in a few months.

### **2. Login page loads flows from default homeserver, pings identity server**

*   `GET https://matrix-client.matrix.org/_matrix/client/versions`
*   `GET https://vector.im/_matrix/identity/api/v1`
*   `GET https://matrix-client.matrix.org/_matrix/client/r0/login`

The login page defaults to immediately presenting the login form for the default homeserver. This means there’s no easy way to login to a custom homeserver from the default configuration of Riot without triggering a request to matrix.org. 

For this section, we have not yet found clear path to removing these requests without degrading the UX for most users. The current status on this section is:

*   Riot makes requests to its default homeserver when users click log in or create account
*   It’s possible that in the future we could discover a UX where we avoid this, but it rests on users having to choose a server, which is something they know nothing about yet 
*   We’ll revisit this at some point in the future
*   For now, you will need to use a custom configuration to avoid these requests

### **3. Application update check**

*   `GET https://riot.im/develop/version`
*   `GET https://packages.riot.im/desktop/update/*`

Riot Web and Desktop automatically check for updates (on web every 10 minutes, on desktop every hour). There is no option to disable the update check, so all default distributions of Riot are making these requests. 

For this section, we believe auto-updating is the best practice for Riot users to ensure they are not stranded on older versions and receive all security updates in a timely fashion. The status for this section is:

* Auto-updating for default Riot distributions will continue as before to ensure security fixes are delivered quickly
* Users of our default distributions will have already contacted these domains once to retrieve the application, so connecting again for updates does not reveal any new information to state actors
* If you want to avoid such connections, you will need to use a custom configuration

### Summary

I hope it is clear that we are taking this feedback seriously. I know the above is perhaps not what everyone here was hoping for, but at least it should give more detail on our thinking as we work out how to best balance these concerns against what we believe is best for most Riot users.

**Comment by sneak** [comment]
Autoupdates are telemetry (and also remote code execution, granting full access to the endpoint encryption keys to any party who can subvert the software distribution server) whether you think it's in a user's best interest to autoupdate or not.

You really need to get opt-in consent for telemetry such as an autoupdate check, otherwise you are leaking users' private usage data without their consent.

**Comment by redmosaic** [comment]
How can I configure linux installation of riot.im desktop client before first launch to avoid unnecessary requests?

**Comment by sneak** [comment]
> For this section, we have not yet found clear path to removing these requests without degrading the UX for most users. 

Do I need to flowchart or pseudocode this for you to show you how to not make requests to a server a user does not ever want to use?

**Comment by auscompgeek** [comment]
@redmosaic https://github.com/vector-im/riot-web/blob/develop/docs/config.md

**Comment by t3chguy** [comment]
https://github.com/vector-im/riot-desktop#user-specified-configjson

**Comment by ara4n** [comment]
@sneak as should be pretty clear from https://github.com/vector-im/riot-web/issues/11655#issuecomment-646065432, we're going to fix this, and yes, autoupdates need an opt-in. We're going to land the first-time user experience work we have in flight first though.

Thanks for bringing it up.

**Comment by sneak** [comment]
I didn't see anything in part 2 of comment 11655 that suggests that the telemetry events described in 2 are going to be fixed, which is why I expressed my disappointment in the "edit the config file" response, which is something that absolutely none of the users I wish to invite to my homeserver will be able to carry out.  If anything, it makes it sound like it's *not* going to be fixed.

**Comment by t3chguy** [comment]
Phase 1 is tracked as https://github.com/vector-im/element-web/issues/11154

**Comment by turt2live** [comment]
For clarity/to summarize, this requires a rework of our entire application startup. Currently it's considered a feature that the app validates the config, however if Product decides otherwise then we can take a look.

I think this requires a major refactoring of our understanding for the application lifecycle, however, and therefore requires a dedicated project.

Placing into the design/product queue for now.

**Comment by 3nprob** [comment]
Just an observation: Validating config is great but connecting to validate if the configured endpoints are available is best seen as a separate step. That is:

`Validation check = Config validation (offline) + services availability check (online)`

As mentioned by @sneak and @ara4n , the offline check can (like today) be done immediately on startup. The latter needs explicit opt-in for both ethical and legal compliance reasons.

Even with a valid config, the configured endpoints can be unavailable or misbehaving for a variety of reasons, which is orthogonal to validating configuration. I see no explicit mention here that the wish from Product to validate config is in conflict with the need for all server interactions to be explicit and opt-in, including auto-updates and prelogin validation.

**Comment by soda-pop-ice-cream** [comment]
So, exactly what settings need to be changed in the `config.json` file? Can you provide example of "silent" `config.json`?

**Comment by t3chguy** [comment]
@soda-pop-ice-cream take your favourite config.json e.g. https://app.element.io/config.json and replace any instances of servers you don't like with ones which you do.

**Comment by soda-pop-ice-cream** [comment]
> [@soda-pop-ice-cream](https://github.com/soda-pop-ice-cream) take your favourite config.json e.g. https://app.element.io/config.json and replace any instances of servers you don't like with ones which you do.

I understand how to use this config with element-web, but what about element-desktop?

**Comment by aaronraimist** [comment]
@soda-pop-ice-cream the path to put the `config.json` at is documented here: https://github.com/element-hq/element-desktop/?tab=readme-ov-file#user-specified-configjson

**Comment by soda-pop-ice-cream** [comment]
TL;DR: You need to create file config.json with content(check path in [aaronraimist](https://github.com/element-hq/element-web/issues/11655#issuecomment-2951248955) comment):

```json
{
  "default_server_config": {
    "m.homeserver": {
      "base_url": "https://your-matrix.server"
    }
  }
}
```

And it works great. In restricted environments where clients have no access to matrix.org or vector.im element-desktop takes 5 minutes to start. After creating this file in `~/.config/Element/config.json` element-desktop starts lightning fast.

**Comment by doits** [comment]
Is there a possibility to set it somewhere globally, not per user? So phoning home can be disabled for all users of the machine, not on a per user basis? (Especially useful if the machine is managed by MDM and users come and go.)

**Comment by t3chguy** [comment]
@doits https://github.com/element-hq/element-web/issues/32351 - though I think nowadays you can specify the config path via envvar, so you could point it at a system-wide file if you can set a system-wide envvar. `ELEMENT_DESKTOP_CONFIG_JSON` as documented in `--help`.

https://github.com/element-hq/element-web/blob/develop/apps/desktop/src/electron-main.ts#L69
